<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.24">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>IOTC Biological and Morphological data collection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="Bio_doc_files/libs/clipboard/clipboard.min.js"></script>
<script src="Bio_doc_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="Bio_doc_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="Bio_doc_files/libs/quarto-html/popper.min.js"></script>
<script src="Bio_doc_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Bio_doc_files/libs/quarto-html/anchor.min.js"></script>
<link href="Bio_doc_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Bio_doc_files/libs/quarto-html/quarto-syntax-highlighting-a37c72dd2dbac68997fcdc15a3622e78.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Bio_doc_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Bio_doc_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Bio_doc_files/libs/bootstrap/bootstrap-bb462d781dde1847d9e3ccf7736099dd.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">IOTC Biological and Morphological data collection</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>Fish biological data are essential for the effective management of marine resources. However, collecting and managing these data remain challenging for many government institutions, despite the extensive scientific research conducted on marine life. Most research activities are privately funded, and the resulting data are privately owned, with confidentiality agreements often established before any work begins. Consequently, although biological data are widely collected across the world’s oceans, they are often either not accessible to national institutions or cannot be shared by those institutions.</p>
<p>In addition to these common issues, there are other limitations associated with the biological data that are collected. As a result, the amount of biological data available to the IOTC Secretariat for data processing and stock assessment is limited. Several working parties have recommended that the Secretariat develop and host a biological database; however, progress has been constrained by the scarcity of accessible information. Recently, the Secretariat has introduced a new reporting form to assist member countries in collecting and submitting biological data.</p>
<p>This paper reviews the biological data currently held by the Secretariat, examines collaborative efforts with partner institutions aimed at developing a comprehensive biological database, and introduces the newly implemented biological data-collection form</p>
</section>
<section id="background" class="level1">
<h1>Background</h1>
<p>Biological data encompass a broad range of information used in scientific analyses, including measurements of morphometric and other biological characteristics of a species. These data are essential for assessing fish populations and understanding stock status. Such information is crucial for supporting management decisions related to fisheries conservation and resource use.</p>
<section id="the-common-biological-data-categorisations" class="level3">
<h3 class="anchored" data-anchor-id="the-common-biological-data-categorisations">The common biological data categorisations:</h3>
<section id="taxonomic-and-specimen-identification" class="level4">
<h4 class="anchored" data-anchor-id="taxonomic-and-specimen-identification">Taxonomic and specimen identification</h4>
<p>This includes the identification of the species, genus/family, sex, and scientific name of the specimen.</p>
</section>
<section id="physical-and-morphometric-data" class="level4">
<h4 class="anchored" data-anchor-id="physical-and-morphometric-data">Physical and morphometric data</h4>
<p>These consist of measurement-based variables such as length, weight, body proportions, and other external physical features of the fish, as well as the condition of the specimen.</p>
</section>
<section id="environmental-data" class="level4">
<h4 class="anchored" data-anchor-id="environmental-data">Environmental data</h4>
<p>These describe the physical characteristics of the aquatic environment from which samples are collected. Relevant parameters may include depth, temperature, and other environmental conditions that influence species distribution and biological processes.</p>
</section>
<section id="tissue-samples" class="level4">
<h4 class="anchored" data-anchor-id="tissue-samples">Tissue samples</h4>
<p>These involve the collection and analysis of tissues such as gonads, liver, blood, and otoliths. Tissue analyses provide valuable information on diet, maturity stage, reproductive status, growth, and other key aspects of the species’ biology.</p>
<p>These components of biological data, although not exhaustive, are essential for the work carried out by the IOTC Secretariat. While physical, morphometric, and environmental data are the most widely used for preliminary analyses, additional biological sampling, such as tissue analysis, is also crucial for comprehensive assessments.</p>
<p>The Secretariat has relied for many years on historical morphometric datasets for major IOTC species to develop conversion factors. The conversion equations published by the Secretariat are widely used by scientists from member countries for a range of analytical purposes.</p>
</section>
</section>
<section id="historical-biological-data" class="level2">
<h2 class="anchored" data-anchor-id="historical-biological-data">Historical biological data</h2>
<p>In the past, several data-collection projects were implemented to gather a range of biological parameters. The predecessor of the IOTC, the <a href="https://www.fao.org/fishery/docs/CDrom/IOTC_Proceedings(1999-2002)/English/documents/miscellaneous/iptp/iptp_man.htm">Indo-Pacific Tuna Programme (IPTP)</a> , conducted multiple in-country sampling programmes during the 1980s and 1990s. These efforts focused largely on small-scale fisheries, particularly line and gillnet fisheries, which are common in coastal countries, and mainly targeted small tuna species.</p>
<p>With the expansion of industrial fisheries from the mid-1990s onward, a number of biological sampling programmes were established to collect data from industrial fleets, by institutes such as, French National Research Institute for Sustainable Development (IRD, formerly ORSTOM) and the Spanish Institute of Oceanography (IEO). The primary source of these data has been observers placed on board vessels for research and monitoring purposes.</p>
</section>
<section id="availability-of-morphometric-data" class="level2">
<h2 class="anchored" data-anchor-id="availability-of-morphometric-data">Availability of Morphometric Data</h2>
<p>Recently, the Secretariat has been collecting and compiling from several published research documents. These data are available for several IOTC species and originate from a variety of research studies. The available conversion equations are in <a href="#Morp">appendix I</a>.</p>
</section>
<section id="challenges-in-collecting-biological-data" class="level2">
<h2 class="anchored" data-anchor-id="challenges-in-collecting-biological-data">Challenges in Collecting Biological Data</h2>
<p>Collecting biological data can be challenging, particularly when on-board sampling procedures are not in place for certain species. This is one of the main reasons for gaps in size-measurement data. In many fisheries, specimens are processed at sea: headed, tailed, gutted, or cut into pieces, before landing. Species within the same family often share similar external features, making accurate identification difficult once they have been processed. This issue is especially evident for billfish species, for which size-measurement data from small-scale and medium-scale fisheries are largely unavailable. Ongoing projects aim to address these challenges, particularly for billfish <span class="citation" data-cites="Darsigan2025">[@Darsigan2025]</span>.</p>
<p>Furthermore, many countries are making effort to review their biological data of the main tuna species caught in their waters, au</p>
</section>
<section id="technological-approaches-to-sampling" class="level2">
<h2 class="anchored" data-anchor-id="technological-approaches-to-sampling">Technological Approaches to Sampling</h2>
<p>In addition to traditional sampling procedures, new technological initiatives are being developed to improve fish data collection. Some institutions are implementing Electronic Monitoring Systems (EMS) capable of sampling individual fish according to pre-defined strategies set by the system <span class="citation" data-cites="Lekunb2022">[@Lekunb2022]</span>.</p>
</section>
<section id="current-development-of-iotc-biological-databases" class="level2">
<h2 class="anchored" data-anchor-id="current-development-of-iotc-biological-databases">Current Development of IOTC Biological Databases</h2>
<p>Several member countries contributed to the Secretariat activities, particularly in implementing scientific projects requested by the Scientific committee. The Secretariat received funding from some member countries through various means. Currently two major projects are under development which will incorporate the development of the biological data bank.Both projects are funded by the European union, either through direct funding or through other associated project.</p>
<p>The main European Union funded, will include a pilot of large regional sampling programme, with the overarching objective of conduct a scoping study which will analyse the potential options for designing a programme that will allow for the coordination of the collection of representative size measurements for both IOTC and bycatch species in the IOTC Area of Competence.</p>
<p>Tasks to be carried out under this activity include:</p>
<ul>
<li><p>review available size-frequency data, individual morphometric data, and associated relationships for both IOTC and bycatch species within the IOTC area of competence, with the aim of identifying data gaps and setting priorities;</p></li>
<li><p>analysis of potential options for designing a regional sampling programme, reviewing the following aspects in particular:</p>
<ul>
<li>the scope of the sampling programme – which CPCs could be involved and where interventions would be most beneficial;</li>
<li>identification of the realistic objectives that the programme can achieve;</li>
<li>identification of priority species for focusing sampling efforts on;</li>
<li>analysis of the logistics that would need to be considered during the development of a region-wide programme;<br>
</li>
<li>analysis of the resources that would be required to develop and maintain a programme;</li>
<li>analysis and development of suitable collection/preservation protocols for samples;<br>
</li>
<li>analysis of suitable data and sampling management considerations including data and sampling hosting (e.g., centralised vs decentralised), maintenance, coordination and data sharing;</li>
<li>consideration of suitable capacity building activities that would benefit the development and running of the sampling programme.</li>
</ul></li>
<li><p>Conduct a pilot phase to be implemented in three CPCs identified during the scoping phase as those where interventions are expected to be most beneficial. Coordinate with those CPCs to set up teams who can go on to train national staff and implement sampling activities for the following data using a three-tiered approach during a pilot phase:</p>
<ul>
<li>Collection of morphometric measurements;</li>
<li>Collection of biological samples for genetic structure analyses; and</li>
<li>Collection of biological samples for epigenetic ageing and/or Close-Kin Mark-Recapture (CKMR), with albacore and/or bigeye tuna as priority species.</li>
</ul></li>
<li><p>The work will likely focus on albacore, yellowfin, and bigeye tuna, as these are the key species for which age and length data are critical to integrated stock assessments. The scoping study should provide guidance on the recommended prioritization and sequencing of work among these species.</p></li>
</ul>
<p>The second project that will improve the biological data of the Secretariat, will be through the South West Indian Ocean Project (SWIOP), aiming at the development of sample storage and management facilities.</p>
<p>This will be to establish a regional Fish Biological Sample Bank (IO-FishBank) to ensure the long-term preservation, traceability, and accessibility of biological materials collected through research surveys, observer programmes, and fisheries monitoring activities in the Indian Ocean. The initiative will strengthen the scientific basis for stock assessment, biodiversity studies, and ecosystem-based management, while promoting collaboration among regional research institutions.</p>
<p>The specific objectives are to:</p>
<ul>
<li>Establish a dedicated facility with controlled cryogenic storage capacity for biological materials</li>
<li>Develop and deploy a digital system for the management of sample metadata and traceability, aligned with FAIR data principles</li>
<li>Implement standard operating procedures for the reception, processing, and cataloguing of samples</li>
<li>Facilitate regional collaboration and capacity development to ensure sustainable use and governance of the repository</li>
</ul>
<p>This task will be composed of four inter-related activities:</p>
<ol type="1">
<li><p>Infrastructure setup and system integration</p>
<ul>
<li>Install one -80 °C freezer (capacity ~24,000 cryotubes) and two -20 °C units for bulk storage, with continuous temperature monitoring and backup power supply</li>
<li>Establish a digital management system integrating a secure database and a web interface for sample metadata, traceability, and access control<br>
</li>
<li>Develop data exchange protocols to ensure interoperability with IOTC and national databases.</li>
</ul></li>
<li><p>Protocol development and standardisation</p></li>
</ol>
<ul>
<li>Prepare and validate standard operating procedures (SOPs) for sample collection, labelling, transport, and archiving, drawing on international best practices (e.g.&nbsp;SPC Pacific Specimen Bank)</li>
<li>Train national and regional partners on SOPs, quality control, and ethical standards.</li>
</ul>
<ol start="3" type="1">
<li><p>Sample collection and catalogue initiation</p>
<ul>
<li>Coordinate with ongoing monitoring and observer programmes to obtain an initial set of priority samples from key IOTC and SIOFA species</li>
<li>Assign unique barcodes and record detailed metadata (species, date, location, tissue type, preservation method, collector, etc.) in the digital catalogue</li>
<li>Implement a pilot workflow covering sample reception, storage, and data entry to test full operational capacity.</li>
</ul></li>
<li><p>Governance, access, and sustainability</p>
<ul>
<li>Establish a regional advisory group to oversee sample access, prioritisation, and benefit-sharing in line with material transfer agreements (MTAs)</li>
<li>Develop an access policy and propose a framework for integration within IOTC’s data collection and research coordination mechanisms</li>
<li>Identify opportunities for co-financing and long-term hosting within an existing marine research centre in the South-West Indian Ocean.</li>
</ul></li>
</ol>
<p>The outputs of the task include:</p>
<ul>
<li>IO-FishBank facility established, equipped with operational cryogenic storage (-80 °C and -20 °C units) and a digital sample management system</li>
<li>Standard operating procedures (SOPs) for sample collection, storage, and metadata management validated and disseminated regionally</li>
<li>Initial catalogue of biological samples (target: up to ~16,000 samples) from selected IOTC and SIOFA species</li>
<li>Regional advisory group and governance framework established, including access policy and draft MTA templates</li>
<li>Training session conducted on sample management protocols and data entry.</li>
<li>Integration framework prepared for interoperability with IOTC data systems and regional monitoring programmes.</li>
</ul>
</section>
<section id="iotc-biological-data-collection-and-reporting" class="level2">
<h2 class="anchored" data-anchor-id="iotc-biological-data-collection-and-reporting">IOTC Biological Data Collection and Reporting</h2>
<p>A key element in biological sampling is the methodology applied. The FAO provides detailed guidance on sampling methods for various purposes, including how to collect and record data. Following a review highlighting deficiencies in species identification, the Secretariat recently conducted two regional workshops on species identification. One module focused specifically on the collection of biological samples, covering sampling methodology, types of samples, and proper sampling techniques.</p>
<p>Recognizing the limited biological data available at the Secretariat and responding to calls for a centralized database, the Secretariat introduced a new data collection form, <a href="https://data.iotc.org/reference/latest/forms/Form-5MB.xlsx">Form 5MB</a> along with <a href="https://data.iotc.org/reference/latest/forms/Form-5MB.html">description</a> for collecting and reporting biological and morphometric data.</p>
</section>
<section id="iotc-data-collection-form-form-5mb" class="level2">
<h2 class="anchored" data-anchor-id="iotc-data-collection-form-form-5mb">IOTC Data Collection Form (Form 5MB)</h2>
<p>In addition to the required metadata, the form is divided into three main sections:</p>
<ol type="1">
<li><p>Sampling: records detailed information about the species sampled:</p>
<ul>
<li>Original source of the data: Main collection point, which may include research institutions, logbook data from fishers, or electronic sources such as cameras.</li>
<li>Sampling protocol used: Specifies whether exhaustive, systematic, or randomized sampling was applied.</li>
<li>Catch estimation: Associated with the specimen.</li>
<li>Data elements: Month, area, fishery type, condition of the specimen, and fate (retained for landing or discarded).</li>
</ul></li>
<li><p>Morphometric Data: captures the physical measurements of the specimens:</p>
<ul>
<li>Length: Type of length measured (if more than one measurement was recorded) and the tool used for measurement.</li>
<li>Weight: Type of weight measurement and the tool used. Note: Emphasis is placed on recording the measurement tools, as different tools can produce different values for the same measurement type.</li>
</ul></li>
<li><p>Tissue Sampling: covers biological samples collected from specimens:</p>
<ul>
<li>Type of tissue collected: e.g., otoliths, gonads, blood.</li>
<li>Storage of samples</li>
<li>Maturity stage</li>
<li>Type of measurement or analysis conducted</li>
<li>Measurement values / mass</li>
</ul></li>
</ol>
</section>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<p>The gap in data collection at the Secretariat is evident from the absence of a centralized biological database. Existing information, primarily morphometric data, is insufficient to establish a comprehensive repository. Furthermore, some datasets are incomplete, lacking key information needed to assess accuracy and quality. Collaboration with other scientific institutions, combined with the introduction of Form 5MB, is expected to enhance the collection and availability of biological data, supporting the development of a dedicated database for biological sampling.</p>
</section>
<section id="Morp" class="level1">
<h1>Appendix I: List of morphometric conversion factors available at the Secretariat</h1>
<p><img src="LW_PARAMS_FT_PW.png" id="fig-LW" class="img-fluid" alt="Length-weight conversion factors collected by the Secretariat"> <img src="LL_PARAMS_FT_LI.png" id="fig-LL" class="img-fluid" alt="Length-length conversion factors collected by the Secretariat"></p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>